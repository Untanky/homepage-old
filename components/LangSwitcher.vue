<template lang="pug">
  .group
    div.p-2.rounded-md.bg-gray-300.inline-block(class="active:rounded-t-0")
      gb-flag.mr-1(:code="currentLocale.flag" size="mini")
      span {{ currentLocale.name }}
    nuxt-link.p-2.rounded-md.bg-gray-300.inline-block(v-for="locale in availableLocales" :key="locale.code" :to="switchLocalePath(locale.code)") 
      gb-flag(:code="locale.flag" size="mini")
      span {{ locale.name }}
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'LangSwitcher',
  computed: {
    currentLocale() {
      if (!this.$i18n.locales) return this.$i18n.fallbackLocale;
      return this.$i18n.locales.find((locale) => (typeof locale === 'string' ? locale : locale.code) === this.$i18n.locale);
    },
    availableLocales() {
      if (!this.$i18n.locales) return this.$i18n.fallbackLocale;
      return this.$i18n.locales.filter((locale) => (typeof locale === 'string' ? locale : locale.code) !== this.$i18n.locale);
    },
  },
});
</script>
