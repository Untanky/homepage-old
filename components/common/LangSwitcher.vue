<template lang="pug">
  .lang-switcher
    select.bg-gray-200.pl-4.pr-6.py-2.rounded-md.border-gray-300.border.appearance-none(@change="changeLocale")
      option(v-for="locale in locales" :key="locale.code" :value="locale.code" :selected="locale.isSelected") {{ `${locale.flag}&nbsp;&nbsp;${locale.name}` }}
</template>

<script lang="ts">
import Vue from 'vue';

type SelectEvent = {
  target: HTMLSelectElement;
}

export default Vue.extend({
  name: 'LangSwitcher',
  methods: {
    changeLocale(event: SelectEvent) {
      this.$i18n.setLocale(event.target
        ? event.target.value
        : this.$i18n.fallbackLocale.toString());
    },
  },
  computed: {
    locales() {
      if (!this.$i18n.locales) return [];
      return this.$i18n.locales.map((locale) => {
        if (typeof locale === 'string') {
          return null;
        }

        return {
          isSelected: locale.code === this.$i18n.locale,
          ...locale,
        };
      });
    },
  },
});
</script>

<style lang="postcss">
.lang-switcher {
  @apply inline;
  @apply relative;
}

.lang-switcher::before {
  content: "â€º";
  @apply block;
  @apply absolute;
  @apply right-3;
  @apply -top-1;
  @apply transform;
  @apply rotate-90;
  @apply text-lg;
}
</style>
