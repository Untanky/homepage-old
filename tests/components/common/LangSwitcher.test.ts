/* eslint-disable no-undef */
// global it, expect
import { mount } from '@vue/test-utils';
import LangSwitcher from '../../../components/common/LangSwitcher.vue';

const locales = [
  {
    code: 'en',
    name: 'English',
    iso: 'en-US',
    flag: 'ðŸ‡ºðŸ‡¸',
    file: 'en.js',
  },
  {
    code: 'de',
    name: 'Deutsch',
    iso: 'de-DE',
    flag: 'ðŸ‡©ðŸ‡ª',
    file: 'de.js',
  },
];

const defaultLocale = 'en';

describe('LangSwitcher', () => {
  it('should run', () => {
    const mocks = {
      $i18n: {
        locales,
        locale: defaultLocale,
      },
    };

    const comp = mount(LangSwitcher, { mocks });
    expect(comp.element).toMatchSnapshot();
  });

  it('should run', () => {
    const mocks = {
      $i18n: {
        locales,
        locale: defaultLocale,
        setLocale: (newLocale: string) => {
          expect(newLocale).toBe('de');
        },
      },
    };

    const comp = mount(LangSwitcher, { mocks });
    expect(comp.element).toMatchSnapshot();

    const select = comp.find('select');
    (<HTMLInputElement>select.element).value = 'de';
    select.trigger('change');

    expect(comp.element).toMatchSnapshot();
  });
});
