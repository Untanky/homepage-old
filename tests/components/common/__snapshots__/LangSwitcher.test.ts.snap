// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LangSwitcher should run 1`] = `
<div
  class="lang-switcher"
>
  <select
    class="bg-gray-200 pl-4 pr-6 py-2 rounded-md border-gray-300 border appearance-none"
  >
    <option
      value="en"
    >
      ğŸ‡ºğŸ‡¸Â Â English
    </option>
    <option
      value="de"
    >
      ğŸ‡©ğŸ‡ªÂ Â Deutsch
    </option>
  </select>
</div>
`;

exports[`LangSwitcher should run 2`] = `
<div
  class="lang-switcher"
>
  <select
    class="bg-gray-200 pl-4 pr-6 py-2 rounded-md border-gray-300 border appearance-none"
  >
    <option
      value="en"
    >
      ğŸ‡ºğŸ‡¸Â Â English
    </option>
    <option
      value="de"
    >
      ğŸ‡©ğŸ‡ªÂ Â Deutsch
    </option>
  </select>
</div>
`;

exports[`LangSwitcher should run 3`] = `
<div
  class="lang-switcher"
>
  <select
    class="bg-gray-200 pl-4 pr-6 py-2 rounded-md border-gray-300 border appearance-none"
  >
    <option
      value="en"
    >
      ğŸ‡ºğŸ‡¸Â Â English
    </option>
    <option
      value="de"
    >
      ğŸ‡©ğŸ‡ªÂ Â Deutsch
    </option>
  </select>
</div>
`;
